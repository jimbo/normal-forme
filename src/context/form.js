import { createContext, useCallback, useContext } from "react"

import ScopeContext from "../context/scope"
import { getDeepValue, join } from "../util/deep"

const FormContext = createContext()
const { Consumer, Provider } = FormContext

export default FormContext
export { Consumer as FormConsumer }
export { Provider as FormProvider }

export const useFormContext = field => {
  const scope = useContext(ScopeContext)
  const scopedField = join(scope, field)
  const [state, dispatch] = useContext(FormContext)
  const value = getDeepValue(state.valueMap, scopedField)

  const setValue = useCallback(
    nextValue =>
      dispatch({
        type: "set value",
        payload: { field: scopedField, nextValue },
      }),
    [dispatch, scopedField]
  )

  const transformValue = useCallback(
    transformValue =>
      dispatch({
        type: "transform value",
        payload: { field: scopedField, transformValue },
      }),
    [dispatch, scopedField]
  )

  return { setValue, transformValue, value }
}
