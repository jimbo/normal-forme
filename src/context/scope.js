import { createContext } from "react"

const ScopeContext = createContext("")
const { Consumer, Provider } = ScopeContext

export default ScopeContext
export { Consumer as ScopeConsumer }
export { Provider as ScopeProvider }
