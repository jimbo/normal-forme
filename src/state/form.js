import { useReducer } from "react"

import { getDeepValue, setDeepValue } from "../util/deep"

const getInitialState = (initialValueMap = new Map()) => ({
  valueMap: initialValueMap,
})

export const reducer = (state = getInitialState(), { payload, type } = {}) => {
  switch (type) {
    case "set value": {
      const { valueMap } = state
      const { field, nextValue } = payload
      const nextValueMap = setDeepValue(valueMap, field, nextValue)

      return { valueMap: nextValueMap }
    }
    case "transform value": {
      const { valueMap } = state
      const { field, transformValue } = payload
      const currentValue = getDeepValue(valueMap, field)
      const nextValue = transformValue(currentValue)
      const nextValueMap = setDeepValue(valueMap, field, nextValue)

      return { valueMap: nextValueMap }
    }
    default: {
      return state
    }
  }
}

export const useFormState = initialValueMap =>
  useReducer(reducer, getInitialState(initialValueMap))
