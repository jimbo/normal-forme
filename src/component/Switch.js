import { createElement, useCallback } from "react"

import { useFormContext } from "../context/form"

const Switch = props => {
  const { field, ...restProps } = props
  const { setValue, value } = useFormContext(field)
  const checked = !!value

  const handleChange = useCallback(
    event => {
      setValue(event.target.checked)
    },
    [setValue]
  )

  return (
    <input
      {...restProps}
      checked={checked}
      onChange={handleChange}
      type="checkbox"
    />
  )
}

export default Switch
