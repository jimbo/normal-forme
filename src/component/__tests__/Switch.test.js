import { createElement } from "react"
import TestRenderer from "react-test-renderer"

import Switch from "../Switch"
import { FormProvider } from "../../context/form"

const field = "a"
const dispatch = jest.fn()
const getState = value => ({ valueMap: new Map().set(field, value) })

describe("Checkbox", () => {
  it("renders correctly", () => {
    const state = getState(true)
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Switch field={field} />
      </FormProvider>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("renders if context value is undefined", () => {
    const state = getState()
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Switch field={field} />
      </FormProvider>
    )

    const input = tree.root.findByType("input")

    expect(input.props.checked).toBe(false)
  })

  it("renders if context value is null", () => {
    const state = getState(null)
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Switch field={field} />
      </FormProvider>
    )

    const input = tree.root.findByType("input")

    expect(input.props.checked).toBe(false)
  })

  it("sets `checked` to false if `value` is false", () => {
    const state = getState(false)
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Switch field={field} />
      </FormProvider>
    )

    const input = tree.root.findByType("input")

    expect(input.props.checked).toBe(false)
  })

  it("sets `checked` to true if `value` is true", () => {
    const state = getState(true)
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Switch field={field} />
      </FormProvider>
    )

    const input = tree.root.findByType("input")

    expect(input.props.checked).toBe(true)
  })

  it("dispatches on change", () => {
    const state = getState(false)
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Switch field={field} />
      </FormProvider>
    )

    const input = tree.root.findByType("input")

    input.props.onChange({
      target: { checked: true },
    })

    expect(dispatch).toHaveBeenCalledTimes(1)
    expect(dispatch).toHaveBeenLastCalledWith({
      payload: { field, nextValue: true },
      type: "set value",
    })
  })
})
