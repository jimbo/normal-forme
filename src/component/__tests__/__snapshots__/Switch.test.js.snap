// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<input
  checked={true}
  onChange={[Function]}
  type="checkbox"
/>
`;
